### PR TITLE
8281632: riscv: Improve interpreter stack banging

### PR DESCRIPTION
**Pull request by yadongw**
With reference to https://bugs.openjdk.java.net/browse/JDK-8072070, there is the same issue on the riscv platform.
 
```
Every time we enter a method in interpreter, we bang a lot of stack ahead for shadow zone checks. I believe this could be made significantly better, e.g. by checking how far away we are from the stack_overflow_limit.
```

Hotspot/jdk tier1 passed on unmatched, and all jtreg cases were tested on Qemu without new failures.

Performed the performance regression test in SPECjvm2008 sunflow, and achieved 17% improvement in the interpreter-only mode.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281632](https://bugs.openjdk.java.net/browse/JDK-8281632): riscv: Improve interpreter stack banging

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to ea9ad368444916afb8c177ba5b29a7e49ac497d5
 * [Fei Yang](https://openjdk.java.net/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/riscv-port pull/58/head:pull/58` \
`$ git checkout pull/58`

Update a local copy of the PR: \
`$ git checkout pull/58` \
`$ git pull https://git.openjdk.java.net/riscv-port pull/58/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 58`

View PR using the GUI difftool: \
`$ git pr show -t 58`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/riscv-port/pull/58.diff">https://git.openjdk.java.net/riscv-port/pull/58.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp [hunk 1]
```diff
@@ -895,16 +895,36 @@ address TemplateInterpreterGenerator::generate_CRC32C_updateBytes_entry(Abstract
...
-    __ sd(zr, Address(t1));
+  const int n_shadow_pages = shadow_zone_size / page_size;
+
+#ifdef ASSERT
+  Label L_good_limit;
+  __ ld(t0, Address(xthread, JavaThread::shadow_zone_safe_limit()));
+  __ bnez(t0, L_good_limit);
+  __ stop("shadow zone safe limit is not initialized");
+  __ bind(L_good_limit);
+#endif
```
Comment (shipilev) [inline src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp:916]: The x86 patch also verifies `shadow_zone_growth_watermark()`, do you want to do the same?

https://github.com/openjdk/jdk/commit/3a13425bc9088cbb6d95e1a46248d7eba27fb1a6#diff-b43598c71e4c69b246ff38275b5d421bcbbc3141216aa03f50fc747dd2635922R742-R746

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp [hunk 2]
```diff
@@ -895,16 +895,36 @@ address TemplateInterpreterGenerator::generate_CRC32C_updateBytes_entry(Abstract
...
+
+  __ ld(t0, Address(xthread, JavaThread::shadow_zone_growth_watermark()));
+  __ bgtu(sp, t0, L_done);
+
+  for (int p = 1; p <= n_shadow_pages ; p++) {
+    __ bang_stack_with_offset(p * page_size);
   }
+
+  // Record a new watermark, unless the update is above the safe limit.
+  // Otherwise, the next time around a check above would pass the safe limit.
```
Comment (shipilev) [inline src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp]: Hah. I noticed this is worded incorrectly in the original x86 change. See https://github.com/openjdk/jdk/pull/7569 -- probably makes sense to wait a bit for that change to land, and then pick up the same wording here?

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp (hunk 1 above)
Comment (yadongw) [reply src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp:916]: Yes, I missed the logic.

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp (hunk 2 above)
Comment (yadongw) [reply src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp]: OK, I've synchronized it, and wait for your patch to be merged.

Comment (shipilev) [reply src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp]: My patch was just merged.

**Review by shipilev — APPROVED** [review]
Looks fine to me.

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp [hunk 3]
```diff
@@ -895,16 +895,42 @@ address TemplateInterpreterGenerator::generate_CRC32C_updateBytes_entry(Abstract
...
+  __ stop("shadow zone growth watermark is not initialized");
+  __ bind(L_good_watermark);
+#endif
+
+  Label L_done;
+
+  __ ld(t0, Address(xthread, JavaThread::shadow_zone_growth_watermark()));
+  __ bgtu(sp, t0, L_done);
+
+  for (int p = 1; p <= n_shadow_pages ; p++) {
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp]: Extra whitespace in loop condition. Looks good otherwise.

**Review by RealFYang — APPROVED** [review]

**Pull request closed without merging**